Fix #7722: Fix warning for guest users

ankita240796 opened the pull request:
## Explanation
Fixes #7722: Fixes warning message for guest users by adding a variable for not checking editing rights if user is not logged in.

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python -m scripts.pre_commit_linter` and `python -m scripts.run_frontend_tests`.
- [x] The PR is made from a branch that's **not** called "develop".
- [ ] The PR has an appropriate "PROJECT: ..." label (Please add this label for the first-pass review of the PR).
- [x] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [x] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

ankita240796 committed:
Fix warning for non-logged in user

ankita240796 committed:
Fix test

DubeySandeep commented on core/controllers/editor.py:
```diff
@@ -780,6 +780,7 @@ class LearnerAnswerInfoHandler(EditorHandler):
     GET_HANDLER_ERROR_RETURN_TYPE = feconf.HANDLER_TYPE_JSON
+    IS_LOGGED_IN_CHECK_REQUIRED = False
 
     @acl_decorators.can_edit_entity
```
I think wer had planned to change the decorator completely, right?

OR change the frontend such that it doesn't make a call to this handler if not logged in.

DubeySandeep replied:
Ignore this comment as I read @anubhavsinha98's comment on the issue thread after this review!

DubeySandeep commented on core/controllers/editor.py:
```diff
@@ -780,6 +780,7 @@ class LearnerAnswerInfoHandler(EditorHandler):
     """Handles the learner answer info for an exploration state."""
 
     GET_HANDLER_ERROR_RETURN_TYPE = feconf.HANDLER_TYPE_JSON
+    IS_LOGGED_IN_CHECK_REQUIRED = False
```
Ah, let's not allow bypassing can_edit_entity decorator via a gust user.

ankita240796 replied:
Updated the code to have changes in frontend instead of the decorator as per your suggestion.

DubeySandeep commented on core/controllers/acl_decorators.py:
```diff
@@ -2599,6 +2599,8 @@ def test_can_edit_entity(self, entity_type, entity_id, **kwargs):
+        if not self.IS_LOGGED_IN_CHECK_REQUIRED and not self.user_id:
+            return reduced_handler(self, entity_id, **kwargs)
         if entity_type == feconf.ENTITY_TYPE_EXPLORATION:
             return can_edit_exploration(reduced_handler)(
```
This will fail for the logged-in user with no edit rights.

ankita240796 replied:
Removed now.

DubeySandeep reviewed (changes requested):
@ankita240796, Thanks for the PR, I've left a few comments belwo PTAL!

Also, going through the issue thread and @anubhavsinha98's comment in the thread I think we need to do the changes in the frontend.

ankita240796 committed:
Make changes to frontend, instead of handler

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into fix-warn

ankita240796 committed:
Fix lint issues

DubeySandeep commented on core/templates/dev/head/pages/exploration-editor-page/services/learner-answer-details-data.service.ts:
```diff
@@ -36,12 +36,23 @@ angular.module('oppia').factory('LearnerAnswerDetailsDataService', [
-          entity_id: _expId});
-
-      return $http.get(learnerAnswerInfoDataUrl);
+      var userIsLoggedIn = false;
```
We have `UserService` which can be helpful here*

ankita240796 replied:
I got this code from the `UserService`. We do not have any function in `UserService` which return user info or info about user being logged in.

ankita240796 replied:
Oh, I got this code from `UserService`. I missed the function which can be called instead. I will make the updates.

ankita240796 committed:
Fix tests

ankita240796 commented:
Hi @DubeySandeep, I have made the suggested changes, PTAL. Thanks!

DubeySandeep commented on core/templates/dev/head/pages/exploration-editor-page/services/learner-answer-details-data.service.ts:
```diff
@@ -36,12 +36,19 @@ angular.module('oppia').factory('LearnerAnswerDetailsDataService', [
+      return $http.get(
+        '/userinfohandler'
+      ).then(function(response) {
+        if (response.data.user_is_logged_in) {
```
I think we want this to work if the user has edit rights, isn't it?

Have you tried testing this with a user who doesn't have the rights to edit the exploration?

ankita240796 replied:
I am sorry I do not get you. If the user is not logged in, we cannot check any rights of the user, right? If the user is logged in the handler will be called as usual and will check the rights as expected.

DubeySandeep replied:
```
Create 2 accounts (A & B):

A creates and published the exploration and B visits the exploration editor page.
```

Have you tried testing the above scenerio^?

ankita240796 replied:
Yes, user B will get an error message:

<img width="1680" alt="Screenshot 2019-10-15 at 11 54 56 PM" src="https://user-images.githubusercontent.com/15226041/66858595-3a684680-efa7-11e9-89fc-a97fefa2ff89.png">

DubeySandeep replied:
So that's not the expected behavior as we allow any user to visit the published exploration editor page!

ankita240796 replied:
Oh Ok, I got it. I have made the changes to fix it. I am using Editability service now to decide whether the user can see the answer info data or not.

ankita240796 committed:
Use UserService

ankita240796 commented on core/templates/dev/head/pages/exploration-editor-page/services/learner-answer-details-data.service.ts:
```diff
@@ -36,12 +37,17 @@ angular.module('oppia').factory('LearnerAnswerDetailsDataService', [
+              entity_id: _expId});
+          return $http.get(learnerAnswerInfoDataUrl);
+        } else {
+          return {data: {learner_answer_info_data: []}};
```
I have a doubt here - Is this the right way to return the data? I get an errror while running the frontend test that the success handler is not called. If I replace this with `$http.get('random')` and add a `$httpBackend.expect` in test and return the data through it, the test work fine. @DubeySandeep can you please guide me on this? Thanks!

ankita240796 replied:
That's strange the frontend tests are passing here but they were failing locally for me.

DubeySandeep replied:
@ankita240796, is this issue resolved?

ankita240796 replied:
No, it isn't. The tests still locally fail when I return the data when the user does not have permission to edit. I return it as follows:

`return $q.resolve({learner_answer_info_data: []})`

The error I get is:

```
Error: Expected spy success to have been called with:
	  [ Object({ learner_answer_info_data: [  ] }) ]
	but it was never called.
```

I just want to confirm once, is there anything wrong with how I am returning the data?

ankita240796 replied:
The frontend tests fail on circleci too.

seanlip reviewed (commented):
One thought -- shouldn't this PR also include some changes to hide the answer directives completely when the user doesn't have access to view them?

seanlip commented:
@anubhavsinha98 please could you take a look at this PR and advise?

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into fix-warn

ankita240796 committed:
Remove warning for users with no edit right

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into fix-warn

ankita240796 committed:
Disable frontend tests

ankita240796 committed:
Enable frontend tests

ankita240796 commented:
Hi @DubeySandeep, sorry for the delay (I was travelling back to home for Diwali). I have addressed the review comments, PTAL. Thanks!

Hi @seanlip, I will look into making the changes as suggested by you once I can get the code for removing the warning work correctly. Thanks!

DubeySandeep commented:
@ankita240796, I see PR #7820 is changing the decorator such that the handler is open for all users so considering that do we need this change?

/cc @brianrodri #7820 will also resolve issue #7722, right?

brianrodri commented:
> @ankita240796, I see PR #7820 is changing the decorator such that the handler is open for all users so considering that do we need this change?
> 
> /cc @brianrodri #7820 will also resolve issue #7722, right?

Yes that's right, we should wait on that PR instead.

ankita240796 commented:
Sounds good, Closing this one then!
